substrate/frame/staking/src/asset.rs

```diff
@@ -41,42 +49,78 @@ pub fn total_balance<T: Config>(who: &T::AccountId) -> BalanceOf<T> {
 ///
 /// This includes balance free to stake along with any balance that is already staked.
 pub fn stakeable_balance<T: Config>(who: &T::AccountId) -> BalanceOf<T> {
-	T::Currency::free_balance(who)
+	free_to_stake::<T>(who).saturating_add(staked::<T>(who))
 }
 
 /// Balance of `who` that is currently at stake.
 ///
-/// The staked amount is locked and cannot be transferred out of `who`s account.
+/// The staked amount is on hold and cannot be transferred out of `who`s account.
 pub fn staked<T: Config>(who: &T::AccountId) -> BalanceOf<T> {
-	T::Currency::balance_locked(crate::STAKING_ID, who)
+	T::Currency::balance_on_hold(&HoldReason::Staking.into(), who)
+}
+
+/// Balance of who that can be staked additionally.
+///
+/// Does not include the current stake.
+pub fn free_to_stake<T: Config>(who: &T::AccountId) -> BalanceOf<T> {
+	T::Currency::balance(who)
 }
 
 /// Set balance that can be staked for `who`.
 ///
-/// This includes any balance that is already staked.
+/// `Value` must be greater than already staked plus existential deposit for free balance.
+///
+/// Should only be used with test.
 #[cfg(any(test, feature = "runtime-benchmarks"))]
 pub fn set_stakeable_balance<T: Config>(who: &T::AccountId, value: BalanceOf<T>) {
-	T::Currency::make_free_balance_be(who, value);
+	use frame_support::traits::fungible::Mutate;
+
+	let staked_balance = staked::<T>(who);
+	// if value is greater than staked balance, we need to increase the free balance.
+	if value > staked_balance {
+		let _ = T::Currency::set_balance(who, value - staked_balance);
+	} else {
+		// else reduce the staked balance.
+		update_stake::<T>(who, value).expect("can remove from what is staked");
+		// burn all free
+		let _ = T::Currency::set_balance(who, Zero::zero());
+	}
+
+	// ensure new stakeable balance same as desired `value`.
+	assert_eq!(stakeable_balance::<T>(who), value);
 }
 
 /// Update `amount` at stake for `who`.
 ///
 /// Overwrites the existing stake amount. If passed amount is lower than the existing stake, the
 /// difference is unlocked.
-pub fn update_stake<T: Config>(who: &T::AccountId, amount: BalanceOf<T>) {
-	T::Currency::set_lock(
-		crate::STAKING_ID,
-		who,
-		amount,
-		frame_support::traits::WithdrawReasons::all(),
-	);
+pub fn update_stake<T: Config>(who: &T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
+	// if first stake, inc provider. This allows us to stake all free balance.
+	if staked::<T>(who) == Zero::zero() && amount > Zero::zero() {
+		frame_system::Pallet::<T>::inc_providers(who);
+	}
```

Comment:
Note that this also means we can't unstake less than ED if there is no free balance.

I see in the function `Ledger::update` that this will result in `NotEnoughFunds` to withdraw. I hope people won't get confused by this.
But it shouldn't happen often.

Or if somebody stake less than ED, then empty his account, then he has to fund the account again to unstake.

Also can this open some spam attack where people can just stake a very small amount send to another account and stake again?

This is only for direct staking no? Is it unreasonable to assume ED is not staked?

Reply:
I think we check and round stake/unstake value to ED if its lower but can verify this again.

> This is only for direct staking no? Is it unreasonable to assume ED is not staked?

Do you mean requiring stakers to maintain ED? Yeah we decided to go with that approach and not have the provider/consumer mess. I will be updating the PR. 

Side effect would be, there are some stakers who have staked all their balance, and we will have to force withdraw 1 DOT from them, which reduces their stake. But in worst case, its roughly 23k DOT reduction in total stake of the system which is not that significant.